drivers/wifi/siwx91x/siwx91x_wifi.c

```diff
@@ -787,6 +787,120 @@ static int siwx91x_dev_init(const struct device *dev)
 	return 0;
 }
 
+static int convert_z_sl_twt_req_type(enum wifi_twt_setup_cmd z_req_cmd)
+{
+	switch (z_req_cmd) {
+	case WIFI_TWT_SETUP_CMD_REQUEST:
+		return REQUEST_TWT;
+	case WIFI_TWT_SETUP_CMD_SUGGEST:
+		return SUGGEST_TWT;
+	case WIFI_TWT_SETUP_CMD_DEMAND:
+		return DEMAND_TWT;
+	default:
+		return -EINVAL;
+	}
+}
+
+static int siwx91x_set_twt(const struct device *dev, struct wifi_twt_params *params)
+{
+	sl_status_t status = SL_STATUS_OK;
+	sl_wifi_twt_request_t twt_req;
+	struct siwx91x_dev *sidev = dev->data;
+	int ret;
+
+	__ASSERT(params, "params cannot be a NULL");
+
+	if (FIELD_GET(SIWX91X_INTERFACE_MASK, sl_wifi_get_default_interface()) !=
+		SL_WIFI_CLIENT_INTERFACE) {
+		params->fail_reason = WIFI_TWT_FAIL_OPERATION_NOT_SUPPORTED;
+		return -ENOTSUP;
+	}
+
+	if (sidev->state != WIFI_STATE_DISCONNECTED && sidev->state != WIFI_STATE_INACTIVE &&
+	    sidev->state != WIFI_STATE_COMPLETED) {
+		LOG_ERR("Command given in invalid state");
+		return -EBUSY;
+	}
+
+	if (params->negotiation_type != WIFI_TWT_INDIVIDUAL) {
+		params->fail_reason = WIFI_TWT_FAIL_OPERATION_NOT_SUPPORTED;
+		return -ENOTSUP;
+	}
+
+	ret = convert_z_sl_twt_req_type(params->setup_cmd);
+	if (ret < 0) {
+		params->fail_reason = WIFI_TWT_FAIL_CMD_EXEC_FAIL;
+		return ret;
+	}
+
+	twt_req.req_type = ret;
+	/*  Device supports Individual TWT only */
+	twt_req.twt_flow_id = params->flow_id;
+	twt_req.negotiation_type = 0;
+
+	if (params->operation == WIFI_TWT_SETUP) {
+		twt_req.twt_enable = 1;
+		twt_req.wake_int_exp = params->setup.twt_exponent;
+		twt_req.triggered_twt = params->setup.trigger;
+		twt_req.un_announced_twt = !params->setup.announce;
+		twt_req.wake_int_mantissa = params->setup.twt_mantissa;
+		twt_req.wake_duration_unit = 0;
+		twt_req.wake_duration = params->setup.twt_wake_interval;
+
+		if (!params->setup.twt_info_disable) {
+			params->fail_reason = WIFI_TWT_FAIL_OPERATION_NOT_SUPPORTED;
+			return -ENOTSUP;
+		}
+		if (TWT_WAKE_INTERVAL_TO_256US(params->setup.twt_wake_interval) > UINT8_MAX) {
+			twt_req.wake_duration_unit = 1;
+			twt_req.wake_duration = TWT_WAKE_INTERVAL_TO_1024US(params->setup.twt_wake_interval);
```

Comment:
`TWT_WAKE_INTERVAL_TO_1024US()` is probably useless. `params->setup.twt_wake_interval / 1024` seems fine.